src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSetSerializer.cs

```diff
@@ -342,6 +371,72 @@ private static Uri GetNestedNextPageLink(ODataSerializerContext writeContext, in
             return null;
         }
 
+        private static Uri GetNestedNextPageLink(ODataSerializerContext writeContext, int pageSize, object obj)
+        {
+            Contract.Assert(writeContext.ExpandedResource != null);
+            IEdmNavigationSource sourceNavigationSource = writeContext.ExpandedResource.NavigationSource;
+            NavigationSourceLinkBuilderAnnotation linkBuilder = writeContext.Model.GetNavigationSourceLinkBuilder(sourceNavigationSource);
+            Uri navigationLink =
+                linkBuilder.BuildNavigationLink(writeContext.ExpandedResource, writeContext.NavigationProperty);
+            Uri nestedNextLink = GenerateQueryFromExpandedItem(writeContext, navigationLink);
+            SkipTokenHandler nextLinkGenerator = null;
+            if (writeContext.QueryOptions != null)
+            {
+                nextLinkGenerator = writeContext.QueryOptions.Context.GetSkipTokenHandler();
+            }
+
+            if (nestedNextLink != null)
+            {
+                if (nextLinkGenerator != null)
+                {
+                    return nextLinkGenerator.GenerateNextPageLink(nestedNextLink, pageSize, obj, writeContext);
+                }
+                return GetNextPageHelper.GetNextPageLink(nestedNextLink, pageSize);
+            }
+
+            return null;
+        }
+
+        private static Uri GenerateQueryFromExpandedItem(ODataSerializerContext writeContext, Uri navigationLink)
+        {
+            IWebApiUrlHelper urlHelper = writeContext.InternalUrlHelper;
+            if (urlHelper == null)
+            {
+                return navigationLink;
+            }
+            string serviceRoot = urlHelper.CreateODataLink(
+                writeContext.InternalRequest.Context.RouteName,
+                writeContext.InternalRequest.PathHandler,
+                new List<ODataPathSegment>());
```

Comment:
If we can get the service root defined for the service, we should use that rather than try and recreate.

Reply:
I was not able to find this saved anywhere. Since this is private, it can be later changed if we start storing it somewhere. 

---
In reply to: [287533506](https://github.com/OData/WebApi/pull/1724#discussion_r287533506) [](ancestors = 287533506)

Reply:
Please create an issue to store this so that we don't have to recreate it here.

---
In reply to: [288310113](https://github.com/OData/WebApi/pull/1724#discussion_r288310113) [](ancestors = 288310113,287533506)